src/approaches/grammar_search_invention_approach.py

```diff
@@ -432,93 +432,124 @@ def learn_from_offline_dataset(self, dataset: Dataset,
+    if CFG.grammar_search_heuristic == "prediction_error":
+        return partial(_prediction_error_heuristic,
+                       initial_predicates,
+                       atom_dataset,
+                       candidates)
+    raise NotImplementedError(
+        f"Unknown heuristic: {CFG.grammar_search_heuristic}.")
+
+
+def _learn_operators_from_atom_dataset(atom_dataset: List[GroundAtomTrajectory]
+        ) -> Tuple[List[Segment], List[STRIPSOperator], List[OptionSpec]]:
+    """Relearn operators and option specs with the given dataset, which may
+    be pruned to include only a subset of the predicates..
+    """
+    segments = [seg for traj in atom_dataset
+                for seg in segment_trajectory(traj)]
+    strips_ops, partitions = learn_strips_operators(segments,
+                                                    verbose=False)
+    option_specs = [p.option_spec for p in partitions]
+    return (segments, strips_ops, option_specs)
+
+
+def _prediction_error_heuristic(initial_predicates: Set[Predicate],
+                                atom_dataset: List[GroundAtomTrajectory],
+                                candidates: Dict[Predicate, float],
+                                s: FrozenSet[Predicate]) -> float:
+    """Score a predicate set by learning operators and counting false positives.
+    """
+    print("Scoring predicates:", s)
+    kept_predicates = s | initial_predicates
+    pruned_atom_data = utils.prune_ground_atom_dataset(atom_dataset,
+                                                       kept_predicates)
+    segments, strips_ops, option_specs = \
+        _learn_operators_from_atom_dataset(pruned_atom_data)
+    num_true_positives, num_false_positives, _, _ = \
+        _count_positives_for_ops(strips_ops, option_specs, segments)
+    # Also add a size penalty.
+    op_size = _get_operators_size(strips_ops)
+    # Also add a penalty based on predicate complexity.
+    pred_complexity = sum(candidates[p] for p in s)
+    total_score = \
+        CFG.grammar_search_false_pos_weight * num_false_positives + \
+        CFG.grammar_search_true_pos_weight * (-num_true_positives) + \
+        CFG.grammar_search_size_weight * op_size + \
+        CFG.grammar_search_pred_complexity_weight * pred_complexity
+    # Useful for debugging:
+    # print("TP/FP/S/C/Total:", num_true_positives, num_false_positives,
+    #       op_size, pred_complexity, total_score)
+    # Lower is better.
+    return total_score
+
+
+def _select_predicates_to_keep(
+        candidates: Dict[Predicate, float],
+        heuristic: Callable[[FrozenSet[Predicate]], float]
+        ) -> Set[Predicate]:
+    """Perform a greedy search over predicate sets.
+    """
+
+    # There are no goal states for this search; run until exhausted.
+    def _check_goal(s: FrozenSet[Predicate]) -> bool:
+        del s  # unused
+        return False
+
+    if CFG.grammar_search_direction == "largetosmall":
```

Comment:
Could we actually just combine these two similar get_successors into one? We can just assign a different lambda function that produces the target we're gonna iterate over (`for predicate in ...`). So like:
```
if CFG.grammar_search_direction == "largetosmall":
    predicates_fn = lambda s: sorted(s)
    init = frozenset(candidates)
else:
    assert CFG.grammar_search_direction == "smalltolarge"
    predicates_fn = lambda s: sorted(set(candidates) - s)
    init = frozenset()
```

and then you only have to define one `_get_successors`

Reply:
Sure can do. Or we can just remove the large to small one because we probably won't use it again anytime soon. Lmk what you prefer

Reply:
Actually for consistency, it would be good to change the `assert` in this code snippet above to a separate `else` condition that raises a NotImplementedError (like we do for the `grammar_search_heuristic` setting), and this can easily be covered via a test that calls `select_predicates_to_keep` with an invalid `grammar_search_direction`.